cabal-install/Distribution/Client/HttpUtils.hs

```diff
@@ -147,14 +146,23 @@ downloadURI verbosity uri path = do
     Left err   -> die $ "Failed to download " ++ show uri ++ " : " ++ show err
     Right rsp -> case rspCode rsp of
       (2,0,0) -> do
-        info verbosity ("Downloaded to " ++ path)
-        writeFileAtomic path $ rspBody rsp
-        return (FileDownloaded path)
+        let index = rspBody rsp
+        case lookupHeader HdrContentLength (rspHeaders rsp) >>= readMaybe of
+            Just contentLength -> do
+              let indexLength = ByteString.length index
+              when (indexLength /= contentLength) $
+                  die $ "Failed to download " ++ show uri
+                        ++ " : Expected content length "   ++ (show contentLength)
+                        ++ " doesn't match actual length " ++ (show indexLength)
+
+              info verbosity ("Downloaded to " ++ path)
+              writeFileAtomic path index 
+              return (FileDownloaded path)
+            Nothing -> die $ "Failed to download " ++ show uri ++ " cannot verify index file length"
```

Comment:
Not sure why you are mentioning the index file here - the function works for any `URI`.